packages/tokens/non-fungible/src/extensions/consecutive/storage.rs

```diff
@@ -425,6 +464,143 @@ impl Consecutive {
+/// significant bit).
+///
+/// # Arguments
+///
+/// * `input` - An optional bitfield (`TokenId`) to search within.
+/// * `start` - Bit index to start the search from, counted from the MSB
+///   (0-based).
+///
+/// # Returns
+///
+/// * `Some(index)` - The index (relative to the MSB) of the first set bit
+///   found, starting from `start` toward the LSB.
+/// * `None` - If `input` is `None`, `start` is out of range, or no set bits are
+///   found.
+///
+/// # Example
+///
+/// If `TokenId::BITS = 8` and `input = Some(0b00101000)`:
+/// - `find_bit_in_item(input, 2)` returns `Some(3)` because the third set bit
+///   (from MSB) is at index 3 (counting from MSB = 0).
+pub(crate) fn find_bit_in_item(input: Option<TokenId>, start: TokenId) -> Option<TokenId> {
+    if let Some(num) = input {
+        let ids_in_item = TokenId::BITS;
+        // Invalid start position
+        if start >= ids_in_item {
+            return None;
+        }
+
+        let last = ids_in_item - 1;
+
+        for i in (0..=(last - start)).rev() {
+            if (num & (1 << i)) != 0 {
+                // i goes from MSB toward LSB relative to `start`, but we want to return
+                // MSB-relative index
+                return Some(last - i);
+            }
         }
     }
+
+    None
+}
+
+/// Searches for the first set bit (`1`) in a vector of bitfields ("bucket"),
+/// starting at a given bit index relative to the entire bucket (from MSB).
+///
+/// Internally, each element in the `bucket` represents a smaller bitfield (an
+/// "item"). This function locates which item to begin with, and then searches
+/// for a set bit across subsequent items using `find_bit_in_item`.
+///
+/// # Arguments
+///
+/// * `bucket` - A vector of `TokenId`s, where each value is treated as a
+///   bitfield.
+/// * `start` - The starting bit index to search from, relative to the MSB of
+///   the whole bucket.
+///
+/// # Returns
+///
+/// * `Some(index)` - The bit index (relative to the entire bucket) of the first
+///   set bit found.
+/// * `None` - If `start` is out of range, or no set bit is found from that
+///   position onwards.
+///
+/// # Example
+///
+/// If `IDS_IN_ITEM = 8`, `ITEMS_IN_BUCKET = 2`, and:
+/// ```
+/// bucket = vec![0b00000000, 0b00101000];
+/// ```
+/// then `find_bit_in_bucket(bucket, 8)` returns `Some(11)`, since bit 3 of the
+/// second item (index 11 in the overall bucket) is the first set bit.
+pub(crate) fn find_bit_in_bucket(bucket: Vec<TokenId>, start: TokenId) -> Option<TokenId> {
+    let ids_in_item = TokenId::BITS;
+    let ids_in_bucket = bucket.len() as TokenId * ids_in_item;
+
+    // Invalid start position
+    if start >= ids_in_bucket {
+        return None;
+    }
+
+    let item_index = start / ids_in_item;
+    // relative id in item
+    let relative_id = start - (item_index * ids_in_item);
+
+    (item_index..bucket.len()).find_map(|i| {
+        let from_id = if i == item_index { relative_id } else { 0 };
```

Comment:
I spent 4 minutes figuring out what does this line do LOL.

If I got it right, it does:
- start from the relative index if we are inspecting the item (so we don't unnecessarily check the previous bits)
- if we are on a later item, then we have to start from `0`

if my understanding is correct, can we put some inline docs here? This whole bitmap thing is probably where our library is working at the most low-level, so I think it deserves some extra inline comments